gpt_neox/data_utils.py

```diff
@@ -42,9 +55,151 @@ def get_tokenizer(tokenizer_type=None, from_pretrained=True, add_padding_token=F
     else:
         raise NotImplementedError('TODO: add custom tokenizers')
 
-def read_enwik8_data(data_path):
-    with gzip.open(data_path) as file:
-        X = np.fromstring(file.read(int(95e6)), dtype=np.uint8)
-        trX, vaX = np.split(X, [int(90e6)])
-        data_train, data_val = torch.from_numpy(trX), torch.from_numpy(vaX)
-    return data_train, data_val
+
+def file_lines(fname):
+    total_lines = 0
+    with open(fname) as f:
+        for l in f:
+            total_lines +=1
+    return total_lines + 1
+
+"""
+Optionally shards data files into a certain size (determined as a multiple of number of entries),
+then creates another "metadata" file that stores a single entry for every entry in the dataset,
+so we can easily index into data for training
+
+shards file is .jsonl with structure:
+{seq_length:INT,total_lines:INT,total_shards:INT}\n
+One line for each index
+{file_name,line}
+...
+"""
+def shardify(data_paths:list,output_path:str, seq_length:int=2048, chunksize:int=None, output_dir:str="",tokenizer=None,num_workers=32):
+    summary_dict = {'seq_length':seq_length,'file_names':[]}
+    
+    shards = []
+    total_shards,total_lines = 0, 0
+    if not os.path.exists(output_dir):
+        os.mkdir(output_dir)
+    for path_idx,path in enumerate(data_paths):
+        total_lines = file_lines(path)
+        num_lines = total_lines//num_workers
+
+        pool = Pool(processes = num_workers)
+        s_process = partial(shardify_process,path=path,num_lines=num_lines,chunksize=chunksize,output_dir=output_dir,\
+            tokenizer=tokenizer,seq_length=seq_length)
+        returns = pool.map(s_process, range(num_workers))
+
+        for line_idx,path_shards,worker_shards,file_names in returns:
+            total_lines += line_idx
+            total_shards += path_shards
+            
+            max_ws = len(summary_dict['file_names'])
+            for ws in worker_shards:
+                ws[0]+=max_ws
+
+            shards.extend(worker_shards)
+            summary_dict['file_names'].extend(file_names)
+                
+    summary_dict['total_lines'] = total_lines
+    summary_dict['total_shards'] = total_shards
+
+    with jsonlines.open(output_dir+"/"+output_path, mode='w') as writer:
+        writer.write(summary_dict)
+        for shard in shards:
+            writer.write(shard)
+
+def shardify_process(worker_id,path,num_lines,chunksize,output_dir,tokenizer,seq_length):
+    start_line = worker_id*num_lines
+    end_line = (worker_id+1)*num_lines
+
+    ext_index = path.find(".")
+    last_slash = path.rfind('/')
+    path_shards,single_file_chunk,single_file_chunk_line = 0,1,0
+        
+    dataset_name = path[last_slash+1:ext_index]
+    extension = path[ext_index:]
```

Comment:
there are much easier and more robust ways to do this with pathlib - if a filename has a dot in anywhere other than the extension for instance, this will break. i'll add changes tomorrow.